### ISSUE TITLE
[MCLEAN-79] Partial Clean + Configurable Selectors

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-clean-plugin#190]
**[Matthew Ouyang](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mouyang)** opened **[MCLEAN-79](https://issues.apache.org/jira/browse/MCLEAN-79?redirect=false)** and commented

Partial cleanup would be useful in situations where source files are deleted or changing between feature branches (which also can result in deleted source files).  This would eliminate the need to do a full clean.

Looking for feedback on provided patch.  To enable this, Selectors need to be configurable + some POC implementations (IdentitySelector, JavaSelector) have between provided.  The patch also includes a page that describes the new feature in more detail, and Selectors have been placed in a separate package (now that there is more than one implementation).

---

**Affects:** 3.0.0

**Attachments:**
- [clean-partial.patch](https://issues.apache.org/jira/secure/attachment/12893437/clean-partial.patch) (_140.59 kB_)

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

IMHO you should avoid using `clean` because, as you say, it removes everything, which is often too much.
For some reason a lot call `mvn clean install` where I highly prefer `mvn verify`.
Plugins should be aware if the task of their goals should be executed or not. e.g. the maven-compiler-plugin verifies nowadays if there are source changes after the latest build. If not, no need to compile.

In other words: it is the plugin responsibiliy to do cleanups when required. The clean goal is just a last solution if plugins don't do their jobs good enough.

**Comment by jira-importer** [comment]
**[Matthew Ouyang](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mouyang)** commented

I didn't realize clean is a separate lifecycle altogether so thank you for pushing me to look into that altogether.  I agree with you that we should be running `verify` instead of `clean install`.

I understand your rationale for having individual plugins handle their own cleanup.  I was worried about code duplication (particularly compiler and resources) and wanted to leverage the directory traversal code, but this seems like a minor worry because it is pretty easy to code a directory traversal.

So what you are saying is that the compiler plugin would remove dangling class files, the resources plugin would remove dangling resources, and any plugin that takes part in the generate-sources phase will remove the sources it generates that later become orphaned.  Finally, clean is the standby in case these plugins don't remove enough files.  Correct?  Please confirm and then I will implement.

Lastly (this goes without saying) don't apply the patch.  At least,, not yet.

**Comment by jira-importer** [comment]
**[Matthew Ouyang](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mouyang)** commented

I commonly run into issues when I am switching branches during development.  Here are some situations I have experienced that my proposal hopes to address.

* My starting branch has a new Java class A.java that the branch I switch to does not have.  Timestamp-based checks won't detect A.java is now gone and delete A.class automatically, and if I don't do a clean then my jar will include A.class and I have experienced cryptic errors later in the pipeline.
* The situation above is also applicable for moving resources from source to target although the consequences are minimal.
* (more obscure) I'm trying to make an SDK compatible with Java 7, and I need to remove new Java 8 features from a class.  If the class was previously compiled in Java 8, it would be nice to recompile this class under Java 7 and timestamp checks won't be sufficient for this either.  Without a clean, I'll eventually run into the "unsupported major.minor version."

It won't be possible to handle every situation (e.g. multiple classes defined in a single source file), but my main objective with this is to make a better efforts towards avoid a `clean`.
